File: test/src/go/t06_mem.go
package main

import (
	"syscall"
	"unsafe"
)

const (
	SYS_CLOSE        = 3
	SYS_MMAP         = 9
	SYS_MPROTECT     = 10
	SYS_MUNMAP       = 11
	SYS_FTRUNCATE    = 77
	SYS_MEMFD_CREATE = 319
)

func main() {
	name := []byte("x\x00")
	fd, _, _ := syscall.RawSyscall(SYS_MEMFD_CREATE, uintptr(unsafe.Pointer(&name[0])), 0, 0)
	_, _, _ = syscall.RawSyscall(SYS_FTRUNCATE, fd, 4096, 0)
	p, _, _ := syscall.RawSyscall6(SYS_MMAP, 0, 4096, 3, 2, fd, 0)
	_, _, _ = syscall.RawSyscall(SYS_MPROTECT, p, 4096, 1)
	_, _, _ = syscall.RawSyscall(SYS_MUNMAP, p, 4096, 0)
	_, _, _ = syscall.RawSyscall(SYS_CLOSE, fd, 0, 0)
}

File: test/src/c/t06_mem.c
#include <unistd.h>
#include <sys/mman.h>
#include <sys/syscall.h>

static int memfd_create_wrap(const char *name, unsigned int flags) {
    return (int)syscall(SYS_memfd_create, name, flags);
}

int main(void) {
    int fd = memfd_create_wrap("x", 0);
    (void)syscall(SYS_ftruncate, fd, 4096);

    void *p = (void*)syscall(SYS_mmap, 0, 4096, 3, 2, fd, 0);
    (void)syscall(SYS_mprotect, p, 4096, 1);
    (void)syscall(SYS_munmap, p, 4096);
    (void)syscall(SYS_close, fd);
    return 0;
}
